service/src/main/java/bio/terra/workspace/service/privateresource/PrivateResourceCleanupService.java

```diff
@@ -0,0 +1,156 @@
+import bio.terra.workspace.db.CronjobDao;
+import bio.terra.workspace.db.ResourceDao;
+import bio.terra.workspace.db.WorkspaceDao;
+import bio.terra.workspace.db.WorkspaceDao.WorkspaceUserPair;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamRethrow;
+import bio.terra.workspace.service.iam.SamService;
+import bio.terra.workspace.service.iam.model.SamConstants.SamResource;
+import bio.terra.workspace.service.iam.model.SamConstants.SamWorkspaceAction;
+import bio.terra.workspace.service.job.JobBuilder;
+import bio.terra.workspace.service.job.JobService;
+import bio.terra.workspace.service.resource.controlled.PrivateResourceState;
+import bio.terra.workspace.service.workspace.flight.RemoveUserFromWorkspaceFlight;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.PostConstruct;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PrivateResourceCleanupService {
+
+  private static final Logger logger = LoggerFactory.getLogger(PrivateResourceCleanupService.class);
+  private static final String PRIVATE_RESOURCE_CLEANUP_JOB_NAME = "private_resource_cleanup_job";
+
+  private final PrivateResourceCleanupConfiguration configuration;
+  private final WorkspaceDao workspaceDao;
+  private final ResourceDao resourceDao;
+  private final CronjobDao cronjobDao;
+  private final SamService samService;
+  private final JobService jobService;
+  private final ScheduledExecutorService scheduler;
+
+  @Autowired
+  public PrivateResourceCleanupService(
+      PrivateResourceCleanupConfiguration configuration,
+      WorkspaceDao workspaceDao,
+      ResourceDao resourceDao,
+      CronjobDao cronjobDao,
+      SamService samService,
+      JobService jobService) {
+    this.configuration = configuration;
+    this.workspaceDao = workspaceDao;
+    this.resourceDao = resourceDao;
+    this.cronjobDao = cronjobDao;
+    this.samService = samService;
+    this.jobService = jobService;
+    this.scheduler = Executors.newScheduledThreadPool(1);
+  }
+
+  @PostConstruct
+  public void startStatusChecking() {
+    if (configuration.isEnabled()) {
+      scheduler.scheduleAtFixedRate(
+          this::cleanupResources,
+          configuration.getStartupWaitSeconds(),
+          TimeUnit.SECONDS.convert(configuration.getPollingIntervalMinutes(), TimeUnit.MINUTES),
+          TimeUnit.SECONDS);
+    }
+  }
+
+  public void cleanupResources() {
+    if (!configuration.isEnabled()) {
+      return;
+    }
+    logger.info("Beginning resource cleanup cronjob");
```

Comment:
My gut feeling is "not very important", since we don't currently have any real constraints on how long this cleanup process should take. 30m feels pretty long, but if it's unacceptably long we could always just set the polling interval to 7.5m instead. Actually running at (polling interval / 2) doesn't really help here unless we undo the claim, but there's always just tweaking the config value.

Per `scheduleAtFixedRate` documentation, "subsequent executions may start late, but will not concurrently execute."

Reply:
I don't want to over-engineer this, especially if there is hope of getting a better trigger from Sam.

I think there are two things I'd like to understand:
- I am still curious if it is safe to run two cleanups concurrently. That seems worth understanding before merging this.
- If the cleanup code fails consistently for some reason (like I insert a bug), would we know? That bothers me about the Stairway flight cleanup also. Unless there is some canonical way to trigger an alert, it could just silently stop working. No fix needed here; just raising the problem.

Reply:
- I don't think it is, running two simultaneous cleanups could lead to clobbering if one flight starts undoing because we don't lock workspaces/resources. They're fine if neither flight starts undoing, though. We have a first pass at mitigation with the cronjob claim table and the fact that within a pod we won't run multiple cleanups, but in a multi-pod environment there is the risk that if one pod's cleanup job takes >15 minutes, another pod will start and we'll have simultaneous flights.
- If the cleanup code fails we'll log an error, but there's no alerting on it at the moment - we'll only see it if someone is checking the logs or error reporting regularly. This seems like a place we might want to consider GCP's log-based alerting, filed PF-1247 to track that.

Reply:
As of the most recent change, it's now safe to run multiple cleanups concurrently. The only remaining risk is if one flight fails after the last step where the flight unlocks resources it's cleaning up and another flight has already grabbed the lock. In that case we'll now log a dismal failure.